tools/preprocess.py

```
@@ -0,0 +1,71 @@
+from PIL import Image, ImageFilter, ImageOps
+
+logger = logging.getLogger("preprocess")
+
+
+class ImagePreprocessor:
+    """Apply simple preprocessing steps to images."""
+
+    AVAILABLE_STEPS = {"grayscale", "binarize", "denoise"}
+
+    def __init__(
+        self, 
```

Ruff (0.11.9)

Trailing whitespace

Remove trailing whitespace

(W291)